multiplatform-movies/ios/Movies/Movies/Gallery/GalleryViewController.swift

```diff
@@ -7,32 +7,73 @@
 
-class GalleryViewController: UIViewController, GalleryPresenterView {
+class GalleryViewController: UICollectionViewController, GalleryPresenterView {
     
     private let presenter = GalleryDependencyProvider(networkingDependencyProvider: NetworkingDependencyProvider()).providerPresenter()
-    private let label = UILabel(frame: CGRect(x: 0, y: 0, width: 300, height: 21))
-
+    private let cellIdentifier = "movie_poster"
+    
+    private var movies : [MoviePoster] = []
+    
     override func viewDidLoad() {
         super.viewDidLoad()
-        label.center = CGPoint(x: 160, y: 285)
-        label.textAlignment = .center
-        label.font = label.font.withSize(25)
-        view.addSubview(label)
+        collectionView?.backgroundColor = .white
+        collectionView?.register(MoviePosterCell.self, forCellWithReuseIdentifier: cellIdentifier)
         presenter.startPresenting(view: self)
     }
     
     func render(gallery: Gallery) {
-        label.text = "Gallery with \(gallery.moviePosters.count) posters"
+        movies = gallery.moviePosters
+        collectionView?.reloadData()
     }
     
     func renderError(message: String?) {
-        guard let errorMessage = message else {
-            label.text = "Something went wrong"
-            return 
-        }
-        label.text = errorMessage
+        print(message)
+    }
+}
+//MARK: - UICollectionViewDatasource
+extension GalleryViewController {
+    
+    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+            return movies.count
+    }
+    
+    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: cellIdentifier, for: indexPath) as! MoviePosterCell
+        let movie = movies[indexPath.row]
+        cell.posterImage.sd_setImage(with: URL(string: movie.thumbnailUrl), placeholderImage: nil)
+        return cell
+    }
+    
+}
+
+extension GalleryViewController: UICollectionViewDelegateFlowLayout {
+    
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        return CGSize(width: collectionView.frame.width/2, height: 310)
     }
 }
 
+class MoviePosterCell: UICollectionViewCell {
+    let posterImage = UIImageView()
+    
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        posterImage.contentMode = .scaleAspectFit
+        posterImage.clipsToBounds = true
+        posterImage.isUserInteractionEnabled = false
+        
+        contentView.addSubview(posterImage)
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        posterImage.frame = contentView.frame
```

Comment:
Instead, we actually wanted to use `AutoLayout`, but couldn't make it work. Could anyone help with that?

Reply:
[Novoda constraints](https://github.com/novoda/novoda-constraints) might help

Reply:
yeah, let's merge this for now so anyone can follow-up on this 👍